src/algorithm/hnswlib/hnswalg.cpp

```diff
@@ -175,47 +175,66 @@ HierarchicalNSW::getDistanceByLabel(LabelType label, const void* data_point) {
 
 bool
 HierarchicalNSW::isValidLabel(LabelType label) {
-    std::unique_lock<std::mutex> lock_table(label_lookup_lock_);
+    std::shared_lock lock_table(label_lookup_lock_);
     bool is_valid = (label_lookup_.find(label) != label_lookup_.end());
     lock_table.unlock();
     return is_valid;
 }
 
+void
+HierarchicalNSW::setBatchNeigohbors(InnerIdType internal_id,
+                                    int level,
+                                    const InnerIdType* neighbors,
+                                    size_t neigbor_count) {
+    linklistsizeint* ll_cur = getLinklistAtLevel(internal_id, level);
+    std::unique_lock lock(link_list_locks_[internal_id]);
+    for (int i = 1; i <= neigbor_count; ++i) {
+        ll_cur[i] = neighbors[i - 1];
+    }
+    setListCount(ll_cur, neigbor_count);
+}
+
+void
+HierarchicalNSW::appendNeigohbor(InnerIdType internal_id,
+                                 int level,
+                                 InnerIdType neighbor,
+                                 size_t max_degree) {
+    linklistsizeint* ll_cur = getLinklistAtLevel(internal_id, level);
+    std::unique_lock lock(link_list_locks_[internal_id]);
+    size_t neigbor_count = getListCount(ll_cur) + 1;
+    if (neigbor_count <= max_degree) {
+        ll_cur[neigbor_count] = neighbor;
+        setListCount(ll_cur, neigbor_count);
+    }
+}
+
 void
 HierarchicalNSW::updateConnections(InnerIdType internal_id,
                                    const vsag::Vector<InnerIdType>& cand_neighbors,
                                    int level,
                                    bool is_update) {
-    linklistsizeint* ll_cur;
-    if (level == 0)
-        ll_cur = getLinklist0(internal_id);
-    else
-        ll_cur = getLinklist(internal_id, level);
+    auto link_data = getLinklistAtLevelWithLock(internal_id, level);
+    linklistsizeint* ll_cur = (linklistsizeint*)link_data.get();
 
     auto cur_size = getListCount(ll_cur);
     auto* data = (InnerIdType*)(ll_cur + 1);
 
-    if (is_update && use_reversed_edges_) {
-        for (int i = 0; i < cur_size; ++i) {
-            auto id = data[i];
-            auto& in_edges = getEdges(id, level);
-            // remove the node that point to the current node
-            std::unique_lock<std::recursive_mutex> lock(link_list_locks_[i]);
-            in_edges.erase(internal_id);
+    if (use_reversed_edges_) {
+        if (is_update) {
+            for (int i = 0; i < cur_size; ++i) {
+                auto id = data[i];
+                auto& in_edges = getEdges(id, level);
```

Comment:
`getEdges` is thread-safe ?

Reply:
This PR only addresses the concurrency issues related to read and write operations in HNSW. The getEdges function will be used during deletion, and the concurrency conflicts in this part will be resolved in a subsequent PR